onnxruntime/core/graph/function_utils.cc

```diff
@@ -0,0 +1,672 @@
+  auto iter = model_local_functions.find(func_identifier);
+  if (iter == model_local_functions.end()) {
+    ORT_THROW("The given function name: ", function_name, ", domain: ", function_domain, " is not found in model local functions");
+  }
+
+  auto* onnx_func_proto = iter->second;
+  ORT_ENFORCE(onnx_func_proto);
+  // generate the schema for this function template
+  // For schema defined functions get the version from the node in parent graph.
+  // For the functions which do not have schema defined (model local functions)
+  // get the since version from the version in opset imports using the domain.
+  auto since_version = GetVersionForDomain(function_domain, domain_version_map);
+  auto op_schema = std::make_unique<ONNX_NAMESPACE::OpSchema>();
+  op_schema->SetName(function_name);
+  op_schema->SetDomain(function_domain);
+  op_schema->SetDoc(onnx_func_proto->doc_string());
+  op_schema->SinceVersion(static_cast<ONNX_NAMESPACE::OperatorSetVersion>(since_version));
+  std::unordered_map<std::string, int> input_name_idx_map;
+  std::unordered_map<std::string, int> output_name_idx_map;
+
+  for (int i = 0; i < onnx_func_proto->input_size(); ++i) {
+    input_name_idx_map[onnx_func_proto->input().Get(i)] = i;
+  }
+  for (int i = 0; i < onnx_func_proto->output_size(); ++i) {
+    output_name_idx_map[onnx_func_proto->output().Get(i)] = i;
+  }
+
+  // Infer a op_schema for stand-alone functions.
+  IOTypeConstraintHelper(*onnx_func_proto, op_schema, input_name_idx_map, output_name_idx_map);
+  auto allow_official_onnx_release_only_final =
+      allow_released_opsets_only && model_load_utils::IsAllowReleasedONNXOpsetsOnlySet();
+
+  const auto onnx_released_versions =
+      schema_registry.GetLastReleasedOpsetVersions(false);
+
+  std::unordered_map<std::string, int> func_domain_to_version;
+  for (auto& opSet : onnx_func_proto->opset_import()) {
+    const auto& domain = opSet.domain();
+    const auto version = gsl::narrow_cast<int>(opSet.version());
+
+    model_load_utils::ValidateOpsetForDomain(onnx_released_versions, logger,
+                                             allow_official_onnx_release_only_final, domain, version);
+
+    // We need to overwrite the domain here with ("") or else the loop below will try to find ("")
+    // in the map and if not found (when domain == kOnnxDomainAlias), adds an entry for ("", 11).
+    // This effectively ignores the opset version specified by the model for the onnx domain.
+    if (domain == kOnnxDomainAlias) {
+      func_domain_to_version[kOnnxDomain] = version;
+    } else {
+      func_domain_to_version[domain] = version;
+    }
+  }
+
+  op_schema->TypeAndShapeInferenceFunction(
+      [onnx_func_proto, func_domain_to_version, &model_local_functions](ONNX_NAMESPACE::InferenceContext& ctx) {
+        auto schema_registry = ONNX_NAMESPACE::OpSchemaRegistry::Instance();
+        ONNX_NAMESPACE::ShapeInferenceOptions options{true, 1, false};
```

Comment:
It means custom schema registry will be broken?

Reply:
what do you mean it is broken?

Reply:
This piece of code only use the global registry.  What if the function needs to use a custom schema?

Reply:
i would dispatch this question to @gramalingam  and @askhade . This PR only did the refactoring work. This question is more about what is the expected behavior in original implementation.

Reply:
Sorry, I am not familiar with this part either. As Cheng says, it is orthogonal to this PR, but it would be good to follow up on this (in a separate PR if necessary). How are custom-registry handled? Can there be multiple opschema registry? If so,  it seems to me even the interfaces will be impacted.

Reply:
Sorry, I don't understand the question. Custom schemas are not a problem, but they also reside in the global registry, right?